internal/controller/controllers/preprovisioningimage_controller.go

```diff
@@ -396,6 +396,104 @@ func (r *PreprovisioningImageReconciler) mapInfraEnvPPI() func(ctx context.Conte
 	return mapInfraEnvPPI
 }
 
+func (r *PreprovisioningImageReconciler) getIronicConfigFromUserOverride(log logrus.FieldLogger, infraEnv *aiv1beta1.InfraEnv) *ICCConfig {
+	ironicAgentImage, ok := infraEnv.GetAnnotations()[ironicAgentImageOverrideAnnotation]
+	if ok && ironicAgentImage != "" {
+		log.Infof("Using override ironic agent image (%s) for infraEnv %s", ironicAgentImage, infraEnv.Name)
+		return &ICCConfig{
+			IronicAgentImage: ironicAgentImage,
+		}
+	}
+
+	return nil
+
+}
+
+func (r *PreprovisioningImageReconciler) getIronicConfigFromBMOConfig(ctx context.Context, log logrus.FieldLogger, infraEnvInternal *common.InfraEnv) *ICCConfig {
+	iccConfig, err := r.BMOUtils.GetICCConfig(ctx)
+	if err != nil {
+		log.WithError(err).Info("ICC configuration is not available")
+		return nil
+	}
+
+	var architectures []string
+	architectures, err = r.OcRelease.GetImageArchitecture(log, iccConfig.IronicAgentImage, infraEnvInternal.PullSecret)
+	if err == nil && funk.Contains(architectures, infraEnvInternal.CPUArchitecture) {
+		log.Infof("Setting ironic agent image (%s) for infraEnv %v from ICC config", iccConfig.IronicAgentImage, infraEnvInternal.Name)
+		return iccConfig
+	}
+
+	log.Infof("CPU architecture (%v) of Ironic agent image (%s) from ICC config is not available for infraEnv %v with arch (%s)",
+		architectures,
+		iccConfig.IronicAgentImage,
+		infraEnvInternal.Name,
+		infraEnvInternal.CPUArchitecture)
+	return nil
+}
+
+func (r *PreprovisioningImageReconciler) getIronicConfigFromHUB(ctx context.Context, log logrus.FieldLogger, infraEnvInternal *common.InfraEnv) *ICCConfig {
+	ironicAgentImage, err := r.getIronicAgentImageByRelease(ctx, log, infraEnvInternal)
```

Comment:
Assuming `getIronicAgentImageByRelease` isn't used anywhere else you can combine these two functions. Not sure there's any reason to have both.

Reply:
merged both methods!